task_manager/storage/asyncpg_storage.py

```diff
@@ -0,0 +1,310 @@
+        }
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(
+            self.create_connection()
+        )
+        loop.run_until_complete(
+            self.__create_schema()
+        )
+        loop.run_until_complete(
+            self.create_subscriber()
+        )
+
+        self.tasks = []  # FIXME what is this ?
+        self.on_task_callbacks: List[OnTaskCallback] = []
+
+    async def __create_schema(self):
+        project_dir = str(Path(
+            os.path.dirname(os.path.realpath(__file__))
+        ).resolve().parents[1])
+
+        sql_migrations_fir_path = os.path.join(
+            project_dir, 'sqlmigrations'
+        )
+        files = []
+
+        # async loop is noway
+        for migration_file in os.listdir(sql_migrations_fir_path):
+            file = open(
+                os.path.join(
+                    sql_migrations_fir_path,
+                    migration_file
+                )
+            )
+            files.append(file.read())
+            file.close()
+
+        async with self.pool.acquire() as connection:
+            connection: Connection
+
+            await connection.execute(
+                "\r".join(files)
+            )
+
+    async def __subscription_callback(
+            self,
+            connection: Connection,
+            pid: int,
+            channel: str,
+            payload: str
+    ):
+        """
+        :param callable callback:
+            A callable or a coroutine function receiving the following
+            arguments:
+            **connection**: a Connection the callback is registered with;
+            **pid**: PID of the Postgres server that sent the notification;
+            **channel**: name of the channel the notification was sent to;
+            **payload**: task idn (str[uuid]).
+        :return:
+        """
+
+        async for callback in generator(self.on_task_callbacks):
+            print(f"callback from postgres: {payload}")
+            await callback(payload)
+
+    async def create_subscriber(self):
+
+        await self.subscribe_conn.add_listener(
+            channel=self.subscription_channel,
+            callback=self.__subscription_callback
+        )
+
+    async def create_connection(self):
+        if not self.pool or self.pool.is_closing():
+            self.pool = await asyncpg.create_pool(
+                **self.connection_data,
+                max_size=10
+            )
+        if not self.subscribe_conn or self.subscribe_conn.is_closed():
+            self.subscribe_conn = await asyncpg.connect(
+                **self.connection_data
+            )
+
+    async def close_connection(self):
+        if self.subscribe_conn:
+            await self.pool.release(self.subscribe_conn)
+            await self.subscribe_conn.close()
+
+        await self.pool.close() if self.pool else None
+
+    async def __create_transaction(self) -> Tuple[Transaction, Connection]:
```

Comment:
because some calls pass a transaction object along the stack. namely TransactionalResult has `commit` and `rollback` methods. In the `AsyncPGConsumedTaskResult` implementation, these methods commit or rollback transactions, respectively.

In this flow, if we use the context manager `async with connection.transaction()`, the transaction will be committed automatically immediately from the output of the context manager and the connection from the pool will be closed, even before calling the `rollback` and `commit` methods from `AsyncPGConsumedTaskResult`

Reply:
https://github.com/requestum-team/task-manager/blob/8c8b6fdad3bd0e4584019716c1b8326b42d5ebaa/task_manager/storage/asyncpg_storage.py#L35-L49

Reply:
we transfer control of the transaction object along the stack to the `TaskManager`

https://github.com/requestum-team/task-manager/blob/8c8b6fdad3bd0e4584019716c1b8326b42d5ebaa/task_manager/core/manager.py#L28-L43

Reply:
that's why we don't use a context manager here - we need to transfer control of the open transaction along the stack to the TaskManager and it then decides to commit it or do a rollback